### PR TITLE
Add documentation for Vectorized environments

### PR DESCRIPTION
**Pull request by tristandeleu**
- Add documentation for vectorized environments (in Sphinx style).
  - Includes documentation to get started with `gym.vector.make`, `AsyncVectorEnv`, and `SyncVectorEnv`
  - Includes documentation about intermediate (shared memory in `AsyncVectorEnv` and exception handling) and advanced (custom spaces) usages of vectorized environments.
  - Includes multiple example snippets.
- Minor fixes in the docstrings of `gym.vector.make`, `VectorEnv`, `AsyncVectorEnv`, and `SyncVectorEnv` for automatic generation of the API reference.

### CODE REVIEW

File: docs/vector/api_reference.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+        :type: :class:`gym.spaces.Space`
+
+        The (batched) action space. The input actions of :meth:`step` must be valid elements of :obj:`action_space`.
+
+        .. code-block::
+
+            >>> envs = gym.vector.make("CartPole-v1", num_envs=3)
+            >>> envs.action_space
+            MultiDiscrete([2 2 2])
```
Comment (vwxyzjn) [inline docs/vector/api_reference.rst:24]: Hmm. What happens when the origin env's action space is `MultiDiscrete`, do you return a tuple of `MultiDiscrete`?

File: docs/vector/getting_started.rst [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    .. code-block::
+
+        >>> envs = gym.vector.AsyncVectorEnv([
+        ...     lambda: gym.make("CartPole-v1"),
+        ...     lambda: gym.make("MountainCar-v0")
+        ... ])
+        RuntimeError: Some environments have an observation space different from `Box([-4.8 ...], [4.8 ...], (4,), float32)`. In order to batch observations, the observation spaces from all environments must be equal.
+
+However, sometimes it may be handy to have access to the observation and action spaces of a sub-environment, and not the batched spaces. You can access those with the properties :attr:`~VectorEnv.single_observation_space` and :attr:`~VectorEnv.single_action_space` of the vectorized environment.
```
Comment (vwxyzjn) [inline docs/vector/getting_started.rst:158]: Thank you!! This is super helpful. I'd suggest highlighting this section with a heading like 

## Working with the spaces of a sub-environment

**Review by vwxyzjn — COMMENTED** [review]
This looks good to me! Could you add another section on recording the episodic returns with the vector env using the following code?

```python
import gym

envs = gym.vector.make("CartPole-v1", num_envs=3)
envs = gym.wrappers.RecordEpisodeStatistics(envs)
envs.reset()
for i in range(100):
    _, _, _, infos = envs.step(envs.action_space.sample())
    for info in infos:
        if "episode" in info.keys():
            print(f"i, episode_reward={info['episode']['r']}")
            break
```

**Comment by jkterry1** [comment]
Hey, I really appreciate all the time you put into this. However, per https://github.com/openai/gym/issues/2276 this isn't compatible with how the documentation website is going to be done. The way to include these docs would be to add elements of this to the api.md docs where applicable, and then to add the remainder to the dev_docs.md page once we have that (or just create it know and the person working on it can add the rest to that file).

**Comment by tristandeleu** [comment]
> However, per #2276 this isn't compatible with how the documentation website is going to be done. The way to include these docs would be to add elements of this to the api.md docs where applicable, and then to add the remainder to the dev_docs.md page once we have that (or just create it know and the person working on it can add the rest to that file).

Is the documentation going to be just a collection of Markdown documents? How do you plan on having cross-references and auto-generated documentation from docstrings?

File: docs/vector/api_reference.rst (hunk 1 above)
Comment (tristandeleu) [reply docs/vector/api_reference.rst:24]: The rules for batching all the standard Gym spaces are available [here](https://github.com/openai/gym/blob/2853ce4797a4dadb1f8702c8e35143d9fbd43ea6/gym/vector/utils/spaces.py#L10). For example a `MultiDiscrete` space becomes a `Box` with an appropriate shape. Do you think it is worth mentioning these conversion rules in the documentation?

**Comment by tristandeleu** [comment]
> Could you add another section on recording the episodic returns with the vector env using the following code?

I think a page on wrappers for vectorized environments (`VectorEnvWrapper` and `Wrapper` compatible with `VectorEnv`) would be very useful indeed, thanks for the suggestion!

**Comment by jkterry1** [comment]
There's a way of hooking the code into the MD files. When a PR for the new Gym website comes in a few weeks you'll see. I'll leave this PR open until then.

**Comment by tristandeleu** [comment]
Any progress on the documentation of vectorized environments? What is the status of the documentation?

**Comment by nikhilweee** [comment]
I was reading up on Vectorised Environments and I happened to stumble upon this PR. Any chance I can view the HTML version of the docs right now? Is there a beta version of the new documentation website that's mentioned in the README?

**Comment by tristandeleu** [comment]
@nikhilweee There is a HTML version of this documentation available here: https://tristandeleu.github.io/gym/vector/

**Comment by nikhilweee** [comment]
Thank you very much, just what I was looking for!

**Pull request merged**